sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -375,15 +375,109 @@ case class GpuSum(child: Expression, resultType: DataType)
     Aggregation.sum().onColumn(inputs.head._2)
 }
 
-case class GpuCount(children: Seq[Expression]) extends GpuDeclarativeAggregate
+/*
+ * GpuPivotFirst is an aggregate function used in the second phase of a two phase pivot to do the
+ * required rearrangement of values into pivoted form.
+ *
+ * For example on an input of
+ * type | A | B
+ * -----+--+--
+ *   b | x | 1
+ *   a | x | 2
+ *   b | y | 3
+ *
+ * with type=groupbyKey, pivotColumn=A, valueColumn=B, and pivotColumnValues=[x,y]
+ *
+ * updateExpressions - In the partial_pivot stage, new columns are created based on
+ * pivotColumnValues one for each of the aggregation. Last aggregation on these columns grouped by
+ * `type` and convert into an array( as per Spark's expectation). Last aggregation(excluding nulls)
+ * works here as there would be atmost one entry in new columns when grouped by `type`.
+ * After CudfLastExcludeNulls, the intermediate result would be
+ *
+ * type | x | y
+ * -----+---+--
+ *   b | 1 | 3
+ *   a | 2 | null
+ *
+ *
+ * mergeExpressions - this is the final pivot aggregation after shuffle. We do another `Last`
+ * aggregation to merge the results. In this example all the data was combined in the
+ * partial_pivot hash aggregation. So it didn't do anything in this stage.
+ *
+ * The final result would be:
+ *
+ * type | x | y
+ * -----+---+--
+ *   b | 1 | 3
+ *   a | 2 | null
+ *
+ * @param pivotColumn column that determines which output position to put valueColumn in.
+ * @param valueColumn the column that is being rearranged.
+ * @param pivotColumnValues the list of pivotColumn values in the order of desired output. Values
+ *                          not listed here will be ignored.
+ */
+case class GpuPivotFirst(
+  pivotColumn: Expression,
+  valueColumn: Expression,
+  pivotColumnValues: Seq[Any],
+  mutableAggBufferOffset: Int = 0,
+  inputAggBufferOffset: Int = 0) extends GpuAggregate {
+
+  val valueDataType = valueColumn.dataType
+
+  override val dataType: DataType = valueDataType
+  override val nullable: Boolean = false
+
+  val pivotColAttr = pivotColumnValues.map(pivotColumnValue => {
+    // If `pivotColumnValue` is null, then create an AttributeReference for null column.
+    if (pivotColumnValue == null) {
+      AttributeReference(GpuLiteral(null, valueDataType).toString, valueDataType)()
+    } else {
+      AttributeReference(pivotColumnValue.toString, valueDataType)()
+    }
+  })
+
+  override lazy val inputProjection: Seq[Expression] = {
+    val expr = pivotColumnValues.map(pivotColumnValue => {
+      if (pivotColumnValue == null) {
+        GpuIf(GpuIsNull(pivotColumn), valueColumn, GpuLiteral(null, valueDataType))
```

Comment:
GpuLiteral does not support arrays, but in the function signature it looks like we support single level arrays. How are we protecting ourselves from valueDataType being an array?

Reply:
I did an explain on a simple query which includes an aggregate(sum) and pivot on a column. At present we dont support arrays for aggregations(sum, max etc). 
The explain of this query produced aggregations on CPU and did a RowToColumnar for pivot. 
For `valueDataType` being array, it throws the exception:
 `class org.apache.spark.sql.catalyst.util.GenericArrayData '[1,1]' is not supported as a scalar yet` .
 I don't think we can support this currently. So changed the function signature. 
 Please let me know if it's okay.

Reply:
You removed the array from the types supported by the Type checks so now this looks fine.